### PR TITLE
feat(workspaces): add support for npm, pnpm, bolt

### PR DESCRIPTION
**Pull request by betaboon**
add support for bolt and pnpm by replacing workspace- and manifest-code with [@manypkgs/get-packages](https://github.com/Thinkmill/manypkg/tree/master/packages/get-packages)

### CODE REVIEW

**Comment by betaboon** [comment]
i think i adapted most tests propertly. the remaining failing one i'm not sure what to do about.

`ReferenceError if path points to a non-file` can be removed imho as that case cannot happen anymore.

`SyntaxError if paths points to package.json with bad syntax` should be converted to provide a workspace to `multiSemanticRelease` and the fixtures should be converted to directories.

**Comment by antongolub** [comment]
@betaboon,

I think we should not just throw away `checkDeps`, `manifest.name` and other checks. Сan we rely on `@manypkg/get-packages` doing all this?

**Comment by betaboon** [comment]
> @betaboon,
> 
> I think we should not just throw away `checkDeps`, `manifest.name` and other checks. Сan we rely on `@manypkg/get-packages` doing all this?

i believe we can rely on `@manypkgs/get-packages` as it uses [`@changesets/types PackageJSON`](https://github.com/atlassian/changesets/blob/master/packages/types/src/index.ts#L43) under the hood.

but to be sure that we can rely on it we could keep the tests in :)

**Comment by antongolub** [comment]
@betaboon, CC @davikawasaki 

Houston, we have a problem. There is an improvement, that brings pkg filtration (https://github.com/qiwi/multi-semantic-release/pull/42). And it hasn't been added to the upstream yet. @betaboon, could you take a look how these mechanics fit with `@manypkg/get-packages` api?

**Comment by davikawasaki** [comment]
@antongolub @betaboon sorry, I've been swamped and didn't send those modifications. Will it make a difference if I commit these features from the fork still today to test with these changes?

**Comment by betaboon** [comment]
> Houston, we have a problem. There is an improvement, that brings pkg filtration ([qiwi#42](https://github.com/qiwi/multi-semantic-release/pull/42)). And it hasn't been added to the upstream yet. @betaboon, could you take a look how these mechanics fit with `@manypkg/get-packages` api?

as i went with a minimal-invasive integrate of `get-packages` it should not be to complicated to integrate that functionality.

@davikawasaki no worries. if you send those changes in a timely fashion i can rework this PR to be compatible.

**Comment by antongolub** [comment]
@betaboon, @davikawasaki,

I suggest to land Davi's improvement at first. As I understand, It was already tested in prod and seems to be ok.

**Comment by betaboon** [comment]
> I suggest to land Davi's improvement at first. As I understand, It was already tested in prod and seems to be ok.

perfectly fine. i am using the changes from this PR by installing the package from a git-branch atm anyway ;)

**Comment by davikawasaki** [comment]
Gotcha, I'll speed up the PR modifications asap. Sorry for the hassle.

**Comment by davikawasaki** [comment]
@antongolub send the PR at https://github.com/dhoulb/multi-semantic-release/pull/53. @betaboon you can see over there the modifications and when approved, try with your code.

**Comment by betaboon** [comment]
> @antongolub send the PR at #53. @betaboon you can see over there the modifications and when approved, try with your code.

on it

**Comment by betaboon** [comment]
@davikawasaki @antongolub done

**Comment by davikawasaki** [comment]
LGTM. @betaboon Can you also provide some fixture tests that ignore packages for bolt?

**Comment by betaboon** [comment]
> LGTM. @betaboon Can you also provide some fixture tests that ignore packages for bolt?

i would have to read up on how bolt handles workspaces. give me some time :)

**Comment by betaboon** [comment]
@davikawasaki  done

**Comment by davikawasaki** [comment]
@betaboon great, looks okay to me. Will it throw an error if no workspaces settings block is set for bolt and pnpm?

**Comment by betaboon** [comment]
> @betaboon great, looks okay to me. Will it throw an error if no workspaces settings block is set for bolt and pnpm?

now it does and now there are tests for it :D

**Comment by davikawasaki** [comment]
> > @betaboon great, looks okay to me. Will it throw an error if no workspaces settings block is set for bolt and pnpm?
> 
> now it does and now there are tests for it :D

Outstanding @betaboon. @antongolub wdyt now?

File: lib/getPackagePaths.js [hunk 1]
```diff
@@ -1,28 +1,38 @@
...
-	if (!checker("string[]+")(packages)) {
-		throw new TypeError("package.json: workspaces or workspaces.packages: Must be non-empty array of string");
+	packages = packages || {
+		tool: "root",
+		root: { dir: cwd },
+	};
+	packages.root.packageJson = getManifest(join(packages.root.dir, "package.json"));
+
+	if (packages.tool === "root") {
+		packages.packages = [];
```
Comment (antongolub) [inline lib/getPackagePaths.js:29]: `packages.packages
` looks a bit confusing. Let's rename to `manypkg.packages` or smth like that?

File: lib/getPackagePaths.js [hunk 2]
```diff
@@ -1,28 +1,38 @@
...
 	}
 
+	// remove cwd from results
+	const cwdRe = new RegExp(`^${cwd + "/"}`);
+	packages = packages.packages.map((p) => p.dir.replace(cwdRe, ""));
```
Comment (antongolub) [inline lib/getPackagePaths.js:34]: How about `path.relative` instead?

**Review by antongolub — CHANGES_REQUESTED** [review]

**Comment by antongolub** [comment]
@betaboon, @davikawasaki

It's almost gltm)

File: lib/getPackagePaths.js (hunk 1 above)
Comment (betaboon) [reply lib/getPackagePaths.js]: in renamed it to `workspace`

File: lib/getPackagePaths.js (hunk 2 above)
Comment (betaboon) [reply lib/getPackagePaths.js]: good point. done :)

**Comment by betaboon** [comment]
> I think we should also mention this new feature in readme.

done

**Comment by davikawasaki** [comment]
> > I think we should also mention this new feature in readme.
> 
> done

@betaboon Make sure you standardize the expected thrown messages at `getPackagePaths.test.js`. Tests failing can be seen [here](https://travis-ci.com/github/dhoulb/multi-semantic-release/jobs/488936416).

**Comment by betaboon** [comment]
> > > I think we should also mention this new feature in readme.
> > 
> > 
> > done
> 
> @betaboon Make sure you standardize the expected thrown messages at `getPackagePaths.test.js`. Tests failing can be seen [here](https://travis-ci.com/github/dhoulb/multi-semantic-release/jobs/488936416).

good catch. should be solved now

**Comment by davikawasaki** [comment]
@betaboon linter problems:

```
/home/travis/build/dhoulb/multi-semantic-release/lib/getPackagePaths.js
  48:32  error  Replace `·` with `⏎↹↹`  prettier/prettier
✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

**Review by antongolub — APPROVED** [review]

**Comment by antongolub** [comment]
@betaboon, LGTM

@davikawasaki, any suggestions / objections?

**Comment by davikawasaki** [comment]
> @betaboon, LGTM
> 
> @davikawasaki, any suggestions / objections?

Nope, feel free to merge.

**Comment by betaboon** [comment]
@antongolub @davikawasaki thank you for all the productive feedback. it's been a treat :)

**Comment by antongolub** [comment]
@betaboon,

Finally merged. Thanks a lot for this improvement and your patience.

**Comment by betaboon** [comment]
you're very welcome.

as always: when i doubt - send patches :D

**Pull request merged**